In [121]:
import re

<link href="style.css" rel="stylesheet"></link>
<article class="day-desc read-aloud"><h2>--- Day 15: Lens Library ---</h2><p>The newly-focused parabolic reflector dish is sending all of the collected light to a point on the side of yet another mountain - the largest mountain on Lava Island. As you approach the mountain, you find that the light is being collected by the wall of a large facility embedded in the mountainside.</p>
<p>You find a door under a large sign that says "Lava Production Facility" and next to a smaller sign that says "Danger - Personal Protective Equipment required beyond this point".</p>
<p>As you step inside, you are immediately greeted by a somewhat panicked <span title="do you like my hard hat">reindeer</span> wearing goggles and a loose-fitting <a href="https://en.wikipedia.org/wiki/Hard_hat" target="_blank">hard hat</a>. The reindeer leads you to a shelf of goggles and hard hats (you quickly find some that fit) and then further into the facility. At one point, you pass a button with a faint snout mark and the label "PUSH FOR HELP". No wonder you were loaded into that <a href="1">trebuchet</a> so quickly!</p>
<p>You pass through a final set of doors surrounded with even more warning signs and into what must be the room that collects all of the light from outside. As you admire the large assortment of lenses available to further focus the light, the reindeer brings you a book titled "Initialization Manual".</p>
<p>"Hello!", the book cheerfully begins, apparently unaware of the concerned reindeer reading over your shoulder. "This procedure will let you bring the Lava Production Facility online - all without burning or melting anything unintended!"</p>
<p>"Before you begin, please be prepared to use the Holiday ASCII String Helper algorithm (appendix 1A)." You turn to appendix 1A. The reindeer leans closer with interest.</p>
<p>The HASH algorithm is a way to turn any <a href="https://en.wikipedia.org/wiki/String_(computer_science)" target="_blank">string</a> of characters into a single <em>number</em> in the range 0 to 255. To run the HASH algorithm on a string, start with a <em>current value</em> of <code>0</code>. Then, for each character in the string starting from the beginning:</p>
<ul>
<li>Determine the <a href="https://en.wikipedia.org/wiki/ASCII#Printable_characters" target="_blank">ASCII code</a> for the current character of the string.</li>
<li>Increase the <em>current value</em> by the ASCII code you just determined.</li>
<li>Set the <em>current value</em> to itself multiplied by <code>17</code>.</li>
<li>Set the <em>current value</em> to the <a href="https://en.wikipedia.org/wiki/Modulo" target="_blank">remainder</a> of dividing itself by <code>256</code>.</li>
</ul>
<p>After following these steps for each character in the string in order, the <em>current value</em> is the output of the HASH algorithm.</p>
<p>So, to find the result of running the HASH algorithm on the string <code>HASH</code>:</p>
<ul>
<li>The <em>current value</em> starts at <code>0</code>.</li>
<li>The first character is <code>H</code>; its ASCII code is <code>72</code>.</li>
<li>The <em>current value</em> increases to <code>72</code>.</li>
<li>The <em>current value</em> is multiplied by <code>17</code> to become <code>1224</code>.</li>
<li>The <em>current value</em> becomes <code><em>200</em></code> (the remainder of <code>1224</code> divided by <code>256</code>).</li>
<li>The next character is <code>A</code>; its ASCII code is <code>65</code>.</li>
<li>The <em>current value</em> increases to <code>265</code>.</li>
<li>The <em>current value</em> is multiplied by <code>17</code> to become <code>4505</code>.</li>
<li>The <em>current value</em> becomes <code><em>153</em></code> (the remainder of <code>4505</code> divided by <code>256</code>).</li>
<li>The next character is <code>S</code>; its ASCII code is <code>83</code>.</li>
<li>The <em>current value</em> increases to <code>236</code>.</li>
<li>The <em>current value</em> is multiplied by <code>17</code> to become <code>4012</code>.</li>
<li>The <em>current value</em> becomes <code><em>172</em></code> (the remainder of <code>4012</code> divided by <code>256</code>).</li>
<li>The next character is <code>H</code>; its ASCII code is <code>72</code>.</li>
<li>The <em>current value</em> increases to <code>244</code>.</li>
<li>The <em>current value</em> is multiplied by <code>17</code> to become <code>4148</code>.</li>
<li>The <em>current value</em> becomes <code><em>52</em></code> (the remainder of <code>4148</code> divided by <code>256</code>).</li>
</ul>
<p>So, the result of running the HASH algorithm on the string <code>HASH</code> is <code><em>52</em></code>.</p>
<p>The <em>initialization sequence</em> (your puzzle input) is a comma-separated list of steps to start the Lava Production Facility. <em>Ignore newline characters</em> when parsing the initialization sequence. To verify that your HASH algorithm is working, the book offers the sum of the result of running the HASH algorithm on each step in the initialization sequence.</p>
<p>For example:</p>
<pre><code>rn=1,cm-,qp=3,cm=2,qp-,pc=4,ot=9,ab=5,pc-,pc=6,ot=7</code></pre>
<p>This initialization sequence specifies 11 individual steps; the result of running the HASH algorithm on each of the steps is as follows:</p>
<ul>
<li><code>rn=1</code> becomes <code><em>30</em></code>.</li>
<li><code>cm-</code> becomes <code><em>253</em></code>.</li>
<li><code>qp=3</code> becomes <code><em>97</em></code>.</li>
<li><code>cm=2</code> becomes <code><em>47</em></code>.</li>
<li><code>qp-</code> becomes <code><em>14</em></code>.</li>
<li><code>pc=4</code> becomes <code><em>180</em></code>.</li>
<li><code>ot=9</code> becomes <code><em>9</em></code>.</li>
<li><code>ab=5</code> becomes <code><em>197</em></code>.</li>
<li><code>pc-</code> becomes <code><em>48</em></code>.</li>
<li><code>pc=6</code> becomes <code><em>214</em></code>.</li>
<li><code>ot=7</code> becomes <code><em>231</em></code>.</li>
</ul>
<p>In this example, the sum of these results is <code><em>1320</em></code>. Unfortunately, the reindeer has stolen the page containing the expected verification number and is currently running around the facility with it excitedly.</p>
<p>Run the HASH algorithm on each step in the initialization sequence. <em>What is the sum of the results?</em> (The initialization sequence is one long line; be careful when copy-pasting it.)</p>
</article>


In [122]:
def HASH_algrothm(s: str) -> int:
    current_value = 0
    for c in s:
        current_value += ord(c)
        current_value *= 17
        current_value %= 256

    return current_value


assert HASH_algrothm("HASH") == 52

tests = [
    ("rn=1", 30),
    ("cm-", 253),
    ("qp=3", 97),
    ("cm=2", 47),
    ("qp-", 14),
    ("pc=4", 180),
    ("ot=9", 9),
    ("ab=5", 197),
    ("pc-", 48),
    ("pc=6", 214),
    ("ot=7", 231),
]

assert all(HASH_algrothm(s) == e for s, e in tests)

In [123]:
steps = "rn=1,cm-,qp=3,cm=2,qp-,pc=4,ot=9,ab=5,pc-,pc=6,ot=7"


def sum_hashes(s: str) -> int:
    return sum(HASH_algrothm(re.sub(r"\n", "", ins)) for ins in s.split(","))


assert sum_hashes(steps) == 1320

In [124]:
with open("./input/day15.txt") as f:
    print(sum_hashes(f.read()))

511215


<link href="style.css" rel="stylesheet"></link>

<main>

<p>Your puzzle answer was <code>511215</code>.</p><p class="day-success">The first half of this puzzle is complete! It provides one gold star: *</p>
<article class="day-desc"><h2 id="part2">--- Part Two ---</h2><p>You convince the reindeer to bring you the page; the page confirms that your HASH algorithm is working.</p>
<p>The book goes on to describe a series of 256 <em>boxes</em> numbered <code>0</code> through <code>255</code>. The boxes are arranged in a line starting from the point where light enters the facility. The boxes have holes that allow light to pass from one box to the next all the way down the line.</p>
<pre><code>      +-----+  +-----+         +-----+
Light | Box |  | Box |   ...   | Box |
-----------------------------------------&gt;
      |  0  |  |  1  |   ...   | 255 |
      +-----+  +-----+         +-----+
</code></pre>
<p>Inside each box, there are several <em>lens slots</em> that will keep a lens correctly positioned to focus light passing through the box. The side of each box has a panel that opens to allow you to insert or remove lenses as necessary.</p>
<p>Along the wall running parallel to the boxes is a large library containing lenses organized by <em>focal length</em> ranging from <code>1</code> through <code>9</code>. The reindeer also brings you a small handheld <a href="https://en.wikipedia.org/wiki/Label_printer" target="_blank">label printer</a>.</p>
<p>The book goes on to explain how to perform each step in the initialization sequence, a process it calls the Holiday ASCII String Helper Manual Arrangement Procedure, or <em>HASHMAP</em> for short.</p>
<p>Each step begins with a sequence of letters that indicate the <em>label</em> of the lens on which the step operates. The result of running the HASH algorithm on the label indicates the correct box for that step.</p>
<p>The label will be immediately followed by a character that indicates the <em>operation</em> to perform: either an equals sign (<code>=</code>) or a dash (<code>-</code>).</p>
<p>If the operation character is a <em>dash</em> (<code>-</code>), go to the relevant box and remove the lens with the given label if it is present in the box. Then, move any remaining lenses as far forward in the box as they can go without changing their order, filling any space made by removing the indicated lens. (If no lens in that box has the given label, nothing happens.)</p>
<p>If the operation character is an <em>equals sign</em> (<code>=</code>), it will be followed by a number indicating the <em>focal length</em> of the lens that needs to go into the relevant box; be sure to use the label maker to mark the lens with the label given in the beginning of the step so you can find it later. There are two possible situations:</p>
<ul>
<li>If there is already a lens in the box with the same label, <em>replace the old lens</em> with the new lens: remove the old lens and put the new lens in its place, not moving any other lenses in the box.</li>
<li>If there is <em>not</em> already a lens in the box with the same label, add the lens to the box immediately behind any lenses already in the box. Don't move any of the other lenses when you do this. If there aren't any lenses in the box, the new lens goes all the way to the front of the box.</li>
</ul>
<p>Here is the contents of every box after each step in the example initialization sequence above:</p>
<pre><code>After "rn=1":
Box 0: [rn 1]

After "cm-":
Box 0: [rn 1]

After "qp=3":
Box 0: [rn 1]
Box 1: [qp 3]

After "cm=2":
Box 0: [rn 1] [cm 2]
Box 1: [qp 3]

After "qp-":
Box 0: [rn 1] [cm 2]

After "pc=4":
Box 0: [rn 1] [cm 2]
Box 3: [pc 4]

After "ot=9":
Box 0: [rn 1] [cm 2]
Box 3: [pc 4] [ot 9]

After "ab=5":
Box 0: [rn 1] [cm 2]
Box 3: [pc 4] [ot 9] [ab 5]

After "pc-":
Box 0: [rn 1] [cm 2]
Box 3: [ot 9] [ab 5]

After "pc=6":
Box 0: [rn 1] [cm 2]
Box 3: [ot 9] [ab 5] [pc 6]

After "ot=7":
Box 0: [rn 1] [cm 2]
Box 3: [ot 7] [ab 5] [pc 6]
</code></pre>

<p>All 256 boxes are always present; only the boxes that contain any lenses are shown here. Within each box, lenses are listed from front to back; each lens is shown as its label and focal length in square brackets.</p>
<p>To confirm that all of the lenses are installed correctly, add up the <em>focusing power</em> of all of the lenses. The focusing power of a single lens is the result of multiplying together:</p>
<ul>
<li>One plus the box number of the lens in question.</li>
<li>The slot number of the lens within the box: <code>1</code> for the first lens, <code>2</code> for the second lens, and so on.</li>
<li>The focal length of the lens.</li>
</ul>
<p>At the end of the above example, the focusing power of each lens is as follows:</p>
<ul>
<li><code>rn</code>: <code>1</code> (box 0) * <code>1</code> (first slot) * <code>1</code> (focal length) = <code><em>1</em></code></li>
<li><code>cm</code>: <code>1</code> (box 0) * <code>2</code> (second slot) * <code>2</code> (focal length) = <code><em>4</em></code></li>
<li><code>ot</code>: <code>4</code> (box 3) * <code>1</code> (first slot) * <code>7</code> (focal length) = <code><em>28</em></code></li>
<li><code>ab</code>: <code>4</code> (box 3) * <code>2</code> (second slot) * <code>5</code> (focal length) = <code><em>40</em></code></li>
<li><code>pc</code>: <code>4</code> (box 3) * <code>3</code> (third slot) * <code>6</code> (focal length) = <code><em>72</em></code></li>
</ul>
<p>So, the above example ends up with a total focusing power of <code><em>145</em></code>.</p>
<p>With the help of an over-enthusiastic reindeer in a hard hat, follow the initialization sequence. <em>What is the focusing power of the resulting lens configuration?</em></p>
</article>
</main>


In [125]:
steps_expected = """
After "rn=1":
Box 0: [rn 1]

After "cm-":
Box 0: [rn 1]

After "qp=3":
Box 0: [rn 1]
Box 1: [qp 3]

After "cm=2":
Box 0: [rn 1] [cm 2]
Box 1: [qp 3]

After "qp-":
Box 0: [rn 1] [cm 2]

After "pc=4":
Box 0: [rn 1] [cm 2]
Box 3: [pc 4]

After "ot=9":
Box 0: [rn 1] [cm 2]
Box 3: [pc 4] [ot 9]

After "ab=5":
Box 0: [rn 1] [cm 2]
Box 3: [pc 4] [ot 9] [ab 5]

After "pc-":
Box 0: [rn 1] [cm 2]
Box 3: [ot 9] [ab 5]

After "pc=6":
Box 0: [rn 1] [cm 2]
Box 3: [ot 9] [ab 5] [pc 6]

After "ot=7":
Box 0: [rn 1] [cm 2]
Box 3: [ot 7] [ab 5] [pc 6]
"""


def parse_box(s: str) -> int:
    return int(re.sub(r'After |"|:', "", s))


def parse_boxes(boxes: list[str]) -> dict[int, str]:
    bs = []
    for box in boxes:
        t = re.sub(r"Box |]", "", box)
        b, c = t.split(":")
        c = re.split(r"\s+\[", c)[1:]
        c = [t.split() for t in c]
        c = {k: int(v) for k, v in c}
        bs.append((int(b), c))
    return dict(bs)


steps_expected = [
    l.splitlines() for l in re.split(r"(?:\r?\n){2,}", steps_expected.strip())
]
steps_expected = [
    (re.sub(r'After |"|:', "", step[0]), parse_boxes(step[1:]))
    for step in steps_expected
]
steps_expected = dict(steps_expected)


actual_boxes = [{} for _ in range(256)]


def perform_step(step: str, boxes: list[list[str]]) -> list[dict[str, int]]:
    m = re.search(r"[=|-]{1}", step)
    op = step[m.start()]
    label, lens = step.split(step[m.start()])
    box = HASH_algrothm(label)

    if op == "=":
        boxes[box][label] = int(lens)
    elif label in boxes[box]:
        del boxes[box][label]
    return boxes


for nr, (step, expected_values) in enumerate(steps_expected.items()):
    actual = perform_step(step, actual_boxes)
    expected = [expected_values.get(i, {}) for i in range(256)]
    assert actual == expected, f"Error in step={step}, nr={nr}"

In [126]:
steps = "rn=1,cm-,qp=3,cm=2,qp-,pc=4,ot=9,ab=5,pc-,pc=6,ot=7"

expected = [steps_expected["ot=7"].get(i, {}) for i in range(256)]


def perfom_steps(steps: str) -> list[dict[str, int]]:
    steps = steps.split(",")
    boxes = [{} for _ in range(256)]
    for step in steps:
        boxes = perform_step(step, boxes)
    return boxes


actual = perfom_steps(steps)
assert actual == expected

In [127]:
steps = "rn=1,cm-,qp=3,cm=2,qp-,pc=4,ot=9,ab=5,pc-,pc=6,ot=7"


def detemine_total_focusing_power(steps: str) -> int:
    boxes = perfom_steps(re.sub(r"\n", "", steps))
    power = 0
    for b, box in enumerate(boxes, start=1):
        if not box:
            continue
        for l, label in enumerate(box, start=1):
            power += b * l * box[label]
    return power


assert detemine_total_focusing_power(steps) == 145

In [128]:
with open("./input/day15.txt") as f:
    print(detemine_total_focusing_power(f.read()))

236057


<link href="style.css" rel="stylesheet"></link>

<main>

<p>Your puzzle answer was <code>236057</code>.</p><p class="day-success">Both parts of this puzzle are complete! They provide two gold stars: **</p>
<p>At this point, you should <a href="/2023">return to your Advent calendar</a> and try another puzzle.</p>
<p>If you still want to see it, you can <a href="15/input" target="_blank">get your puzzle input</a>.</p>
<p>You can also <span class="share">[Share<span class="share-content">on
  <a href="https://twitter.com/intent/tweet?text=I%27ve+completed+%22Lens+Library%22+%2D+Day+15+%2D+Advent+of+Code+2023&amp;url=https%3A%2F%2Fadventofcode%2Ecom%2F2023%2Fday%2F15&amp;related=ericwastl&amp;hashtags=AdventOfCode" target="_blank">Twitter</a>
  <a href="javascript:void(0);" onclick="var ms; try{ms=localStorage.getItem('mastodon.server')}finally{} if(typeof ms!=='string')ms=''; ms=prompt('Mastodon Server?',ms); if(typeof ms==='string' &amp;&amp; ms.length){this.href='https://'+ms+'/share?text=I%27ve+completed+%22Lens+Library%22+%2D+Day+15+%2D+Advent+of+Code+2023+%23AdventOfCode+https%3A%2F%2Fadventofcode%2Ecom%2F2023%2Fday%2F15';try{localStorage.setItem('mastodon.server',ms);}finally{}}else{return false;}" target="_blank">Mastodon</a></span>]</span> this puzzle.</p>
</main>
